src/Interpreters/Cache/FileCache.cpp

```diff
@@ -653,10 +650,11 @@ bool FileCache::tryReserve(FileSegment & file_segment, size_t size)
     auto is_main_priority_overflow = [&]
     {
         /// max_size == 0 means unlimited cache size,
-        /// max_element_size means unlimited number of cache elements.
+        /// max_element_size == 0 means unlimited number of cache elements.
         const bool is_overflow = (main_priority->getSizeLimit() != 0
-                                  && main_priority->getSize(cache_lock) + size - removed_size > main_priority->getSizeLimit())
-            || (main_priority->getElementsLimit() != 0 && queue_size > main_priority->getElementsLimit());
+                                  && (main_priority->getSize(cache_lock) + size - freeable_space > main_priority->getSizeLimit()))
+            || (main_priority->getElementsLimit() != 0
+                && (main_priority->getElementsCount(cache_lock) + 1 - freeable_count > main_priority->getElementsLimit()));
```

Comment:
Please, let's simplify the condition, and probably make it more clean in separate PR (as discussed)
```
freeable_count == 0 && main_priority->getElementsCount() == main_prioirty->getElementsLimit()
```